packages/e2e-tests/specs/editor/various/rich-text.test.js

```diff
@@ -382,39 +405,42 @@ describe( 'RichText', () => {
 		// Add text and select to color.
 		await page.keyboard.type( '1' );
 		await pressKeyWithModifier( 'primary', 'a' );
-		await clickBlockToolbarButton( 'More' );
 
-		const button = await page.waitForXPath(
-			`//button[contains(text(), 'Text color')]`
-		);
-		// Clicks may fail if the button is out of view. Assure it is before click.
-		await button.evaluate( ( element ) => element.scrollIntoView() );
-		await button.click();
+		// Open the color menu.
+		await clickBlockToolbarButton( 'More' );
+		await page.click( 'button:text-is( "Text color" )' );
 
-		// Select color other than black.
-		await page.keyboard.press( 'Tab' );
-		await page.keyboard.press( 'Enter' );
+		// Apply the "Pale pink" color.
+		await page.click( 'button[ aria-label="Color: Pale pink" ]' );
 
+		// Make sure the colored text is visible
+		const coloredTextSelector = '.has-pale-pink-color:text-is( "1" )';
+		await page.waitForSelector( coloredTextSelector, { state: 'visible' } );
 		expect( await getEditedPostContent() ).toMatchSnapshot();
 
 		// Dismiss color picker popover
-		await page.keyboard.press( 'Escape' );
+		await page.keyboard.press( 'Escape', { delay: 100 } );
 
 		// Navigate to the block.
-		await page.keyboard.press( 'Tab' );
+		await page.keyboard.press( 'Tab', { delay: 100 } );
 
 		// Copy the colored text.
 		await pressKeyWithModifier( 'primary', 'c' );
 
 		// Collapsed the selection to the end.
-		await page.keyboard.press( 'ArrowRight' );
+		await page.keyboard.press( 'ArrowRight', { delay: 100 } );
 
 		// Create a new paragraph.
-		await page.keyboard.press( 'Enter' );
+		await page.keyboard.press( 'Enter', { delay: 100 } );
```

Comment:
Nice catch - this is an interesting one! 😄 When running the `rich-text` suite locally, in places where a lot of editor actions were performed in a row and the changes were expected immediately, like in [this test](https://github.com/WordPress/gutenberg/blob/2c924df03a45943eb5e21113b7295436e2facf90/packages/e2e-tests/specs/editor/various/rich-text.test.js#L433-L453), the UI was sometimes having trouble keeping up and thus failing somewhere in the middle, or returning false-negative results. IMO there are two ways of dealing with such situations:

1. Wait for the expected changes (instead of expecting them immediately).
   
   For example, instead of this:

   ```js
	await activeParagraph.type( '1' );
	await page.keyboard.press( 'Enter' ); // Create paragraph.
	await page.keyboard.press( 'Backspace' ); // Remove paragraph.
	await page.keyboard.type( '2' );

	expect( await activeParagraph.innerText() ).toBe( '12' );
   ```

   ...do this, as it ensures the editor actions were performed and the test result is not false-negative (i.e. how do we know the paragraph was even created?):

   ```js
	await activeParagraph.type( '1' );
	await page.keyboard.press( 'Enter' );
	// Make sure the paragraph is created between keypresses...
	const paragraph = await page.waitForSelector( 'p[aria-label*="Empty block"]'	);
	await page.keyboard.press( 'Backspace' );
	// ...and make sure it's removed.
	await paragraph.waitForElementState( 'hidden' );
	await page.keyboard.type( '2' );

	expect( await activeParagraph.innerText() ).toBe( '12' );
   ```

2. Simulate an actual user typing, which means add a delay between `keyup` and `keydown` events (~100ms for an average user):

   ```js
	await activeParagraph.type( 'foo' ); // No delay needed as it doesn't change the DOM structure.
	await page.keyboard.press( 'Enter', { delay: 100 } ); // Create paragraph.
	await page.keyboard.press( 'Backspace', { delay: 100 } ); // Remove paragraph.
	await page.keyboard.type( 'bar' );

	expect( await activeParagraph.innerText() ).toBe( 'foobar' );
   ```

In this situation, I added the delay because it prevents the flakiness, but also is a good indicator of whether the canvas is responsive enough. If the expected result is not produced for an average-speed-typing user, there's something wrong with the app. The obvious downside of this solution is that it can add a lot of seconds if used extensively, so probably the first method would be a preferred one. What do you think?

Reply:
I prefer the first approach. We can never be sure a delay is enough for the assertion. It's more or less the same as `page.waitForTimeout` which is considered an anti-pattern.

As mentioned before in another comment, the ideal solution for this is to have auto-wait assertions for selecting elements. This could be done in `e2e-test-utils` or in a different library.

Reply:
Yeah, the first approach is more solid. We have performance tests to monitor canvas responsivity; here we're just testing that pressing those keys achieves the expected result. Connection speed and device capabilities vary a lot, and CI servers are usually slow, so it's best not to assume any kind of default lag.